omnibus/files/private-chef-cookbooks/private-chef/libraries/private_chef.rb

```diff
@@ -372,7 +306,76 @@ def gen_keepalived(node_name)
       PrivateChef["nginx"]["ha"] ||= true
     end
 
+    #
+    # Genereric gen_ callbacks
+    #
+    # - gen_frontend: Run on all frontend nodes
+    #
+    # - gen_backend: Runs on all backend nodes, take a parameter
+    # indicating whether the current node is the bootstrap node.
+    #
+    # - gen_api_fqdn: Runs on all nodes
+    #
+    # - gen_secrets: Runs on all nodes but should abort if a needed
+    # secret doesn't exist on a non-bootstrap node. Takes the node
+    # name.
+    #
+    # If a plugin doesn't define a callback for one of these functions
+    # the generic version is used.
+    #
+    def gen_frontend
+      callback = callback_for(:gen_frontend)
+      if ! callback.nil?
+        instance_exec(&callback)
+      else
+        gen_frontend_default
+      end
+    end
+
     def gen_backend(bootstrap=false)
+      callback = callback_for(:gen_backend)
+      if ! callback.nil?
+        instance_exec(bootstrap, &callback)
+      else
+        gen_backend_default(bootstrap)
+      end
+    end
+
+    def gen_secrets(node_name)
+      callback = callback_for(:gen_secrets)
+      if ! callback.nil?
+        instance_exec(node_name, &callback)
+      else
+        gen_secrets_default(node_name)
+      end
+    end
+
+    def gen_api_fqdn
+      callback = callback_for(:gen_api_fqdn)
+      if ! callback.nil?
+        instance_exec(&callback)
+      else
+        gen_api_fqdn_default
+      end
+    end
+
+    def callback_for(name)
+      extension = PrivateChef["registered_extensions"][PrivateChef["topology"]]
```

Comment:
I think the answer to this goes hand in hand with whether we end up with plugins for anything other than ha topologies.   I could see us eventually allowing add-ons such as manage to share the chef-server configuration file, but in that case, a lot of this is going to need to change.